libbpf-tools/cpudist.bpf.c

```diff
@@ -0,0 +1,121 @@
+// SPDX-License-Identifier: GPL-2.0
+// Copyright (c) 2020 Wenbo Zhang
+#include "vmlinux.h"
+#include <bpf/bpf_helpers.h>
+#include <bpf/bpf_core_read.h>
+#include <bpf/bpf_tracing.h>
+#include "cpudist.h"
+
+#define TASK_RUNNING 0
+
+const volatile bool targ_per_process = false;
+const volatile bool targ_per_thread = false;
+const volatile bool targ_offcpu = false;
+const volatile bool targ_ms = false;
+const volatile pid_t targ_tgid = -1;
+
+struct {
+	__uint(type, BPF_MAP_TYPE_HASH);
+	__type(key, u32);
+	__type(value, u64);
+} start SEC(".maps");
+
+static struct hist initial_hist;
+
+struct {
+	__uint(type, BPF_MAP_TYPE_HASH);
+	__type(key, u32);
+	__type(value, struct hist);
+} hists SEC(".maps");
+
+static __always_inline u32 log2(u32 v)
+{
+	u32 shift;
+	u32 r;
+
+	r = (v > 0xFFFF) << 4; v >>= r;
+	shift = (v > 0xFF) << 3; v >>= shift; r |= shift;
+	shift = (v > 0xF) << 2; v >>= shift; r |= shift;
+	shift = (v > 0x3) << 1; v >>= shift; r |= shift;
+	r |= (v >> 1);
+
+	return r;
+}
+
+static __always_inline u32 log2l(u64 v)
+{
+	u32 hi = v >> 32;
+
+	if (hi)
+		return log2(hi) + 32 + 1;
+	else
+		return log2(v) + 1;
+}
+
+static __always_inline void store_start(u32 tgid, u32 pid, u64 ts)
+{
+	if (targ_tgid != -1 && targ_tgid != tgid)
+		return;
+	bpf_map_update_elem(&start, &pid, &ts, 0);
+}
+
+static __always_inline void update_hist(u32 tgid, u32 pid, u64 ts)
+{
+	struct hist *histp;
+	u64 delta, *tsp;
+	u32 loc, id;
+
+	if (targ_tgid != -1 && targ_tgid != tgid)
+		return;
+
+	tsp = bpf_map_lookup_elem(&start, &pid);
+	if (!tsp || ts < *tsp)
+		return;
+
+	if (targ_per_process)
+		id = tgid;
+	else if (targ_per_thread)
+		id = pid;
+	else
+		id = -1;
+	histp = bpf_map_lookup_elem(&hists, &id);
+	if (!histp) {
+		bpf_map_update_elem(&hists, &id, &initial_hist, 0);
+		histp = bpf_map_lookup_elem(&hists, &id);
+		if (!histp)
+			return;
+	}
+	delta = ts - *tsp;
+	if (targ_ms)
+		delta /= 1000000;
+	else
+		delta /= 1000;
+	loc = log2l(delta);
+	if (loc >= MAX_SLOTS)
+		return;
+	histp->slots[loc] += 1;
```

Comment:
log2l is always >=1, so we are wasting slot 0, plus putting 0 into slot #1, instead of slot #0. So `loc-1` here?

Reply:
Fixed, thanks.

Now I write as this:

```c
loc = log2l(delta);
if (loc > MAX_SLOTS)
        loc = MAX_SLOTS;
loc -= 1;
if (loc >= MAX_SLOTS)
        return;
histp->slots[loc] += 1;
```

I found without check again with

```c 
if (loc >= MAX_SLOTS)
        return;
```

The verifier will reject:
> R1 unbounded memory access, make sure to bounds check any array access into a map

Reply:
verifier thinks it underflows. You have to `-= 1` first, then check against MAX_SLOTS

Reply:
there shouldn't be `if (loc >= MAX_SLOTS) return;` at all:

```
loc = log2l(delta) - 1; /* or maybe just remove + 1 from log2l? */
if (loc >= MAX_SLOTS) loc = MAX_SLOTS - 1;
histp->slots[loc]++;
```

btw, why `loc` and not `slot` or something like that?

Reply:
> there shouldn't be `if (loc >= MAX_SLOTS) return;` at all:
> 
> ```
> loc = log2l(delta) - 1; /* or maybe just remove + 1 from log2l? */
> if (loc >= MAX_SLOTS) loc = MAX_SLOTS - 1;
> histp->slots[loc]++;
> ```

Thanks @anakryiko , I tired this before, but still get error:

> ; histp->slots[loc]++;
> 345: (61) r1 = *(u32 *)(r6 +0)
>  R0=inv(id=0,umax_value=16,var_off=(0x0; 0x1f)) R1_w=invP(id=0,umax_value=17179869180,var_off=(0x0; 0x3fffffffc)) R2=inv(id=0,umax_value=34,var_off=(0x0; 0x3f)) R3=inv35 R4=inv0 R6_w=map_value(id=0,off=0,ks=4,vs=160,umax_value=17179869180,var_off=(0x0; 0x3fffffffc)) R7=inv(id=0,umax_value=4294967295,var_off=(0x0; 0xffffffff)) R8=inv(id=0,umax_value=4294967295,var_off=(0x0; 0xffffffff)) R9=inv(id=0) R10=fp0 fp-8=mmmm???? fp-16=mmmmmmmm fp-24=mmmmmmmm fp-32=mmmmmmmm fp-40=mmmmmmmm fp-48=map_value
> R6 unbounded memory access, make sure to bounds check any array access into a map
> processed 154 insns (limit 1000000) max_states_per_insn 0 total_states 12 peak_states 12 mark_read 9

It seems `R2=inv(id=0,umax_value=34,var_off=(0x0; 0x3f))` is limited up to 34, but access array with R6.

> 
> btw, why `loc` and not `slot` or something like that?
Sure I'll change to `slot`, the reason is I used to think `slots[slot]` maybe looks not good.

Reply:
Ok, so I played with this locally. The problem is that log2 and log2l returns u32, and slot itself is u32, but then it's used as an index into array, which is used as u64. So Clang generates bit extension:

```
r6 <<= 32
r6 >>= 32
```
which makes verifier lose track of the fact that `slot` is actually < 36. If you change `slot` type to u64 and log2/log2l return type to u64, it will work.

Reply:
Got it, thanks. Fixed.

Reply:
> Ok, so I played with this locally. The problem is that log2 and log2l returns u32, and slot itself is u32, but then it's used as an index into array, which is used as u64. So Clang generates bit extension:
> 
> 
> 
> ```
> 
> r6 <<= 32
> 
> r6 >>= 32
> 
> ```
> 
> which makes verifier lose track of the fact that `slot` is actually < 36. If you change `slot` type to u64 and log2/log2l return type to u64, it will work.

Hi @anakryiko I just found LLVM-7.0 support alu32 to use 32bit subregister, then I tested Makefile with `$(Q)$(CLANG) -g -O2 -target bpf -Xclang -target-feature -Xclang +alu32` and change `log2(l)` return val’s, slot’s type back to `u32`,  found it’s ok with it. So is it better to change to `-Xclang -target-feature -Xclang +alu32`?

Reply:
There were some pretty recent fixes to verifier related to use of alu32. If we switch libbpf-tools to alu32, they might not run successfully on as wide range of kernels as currently. So I'd defer it for now and stick to no-alu32.

Reply:
Get it, thanks!